### PR TITLE
Theano cudnn code now throws Exception when it is not available, need…

### PR DESCRIPTION
**Pull request by jrao1**
… to catch this

See https://github.com/fchollet/keras/issues/4821 for an example of the issue.

### CODE REVIEW

**Comment by fchollet** [comment]
Since when?

**Comment by patyork** [comment]
@fchollet I looked into this a bit. The issue seems to be that CuDNN is [*assumed* installed if CUDA is available and a GPU is being used](https://github.com/fchollet/keras/blob/master/keras/backend/theano_backend.py#L540).

This PR doesn't seem to do much (just ignores the error and doesn't apply BatchNorm) but I think that either that line I linked to above should change, or there should be a try/except to fall back to the non-cudnn batch_norm that T.nnet provides.

**Comment by jrao1** [comment]
@fchollet The commit that introduced the dnn_version() function that throws Exception is https://github.com/Theano/Theano/commit/adb02ace571898b7faef8b61d9792547af8e0dda#diff-66f28a82bf858d82b940e8fa182abe69

**Comment by gvtulder** [comment]
@jrao1: Looks good. (Except for the formatting error in the line I suggested, sorry for that.)

File: keras/backend/theano_backend.py [hunk 1]
```diff
@@ -472,7 +472,8 @@ def _old_normalize_batch_in_training(x, gamma, beta,
     '''Computes mean and std for batch then apply batch_normalization on batch.
     '''
     dev = theano.config.device
-    use_cudnn = ndim(x) < 5 and reduction_axes == [0, 2, 3] and (dev.startswith('cuda') or dev.startswith('gpu'))
+    use_cudnn = ndim(x) < 5 and reduction_axes == [0, 2, 3] and (dev.startswith('cuda') or dev.startswith('gpu')) and \
+        theano.sandbox.cuda.dnn_available()
```
Comment (nouiz) [inline keras/backend/theano_backend.py]: this is not good I think with the new back-end. I think it should be:

    use_cudnn = ndim(x) < 5 and reduction_axes == [0, 2, 3]
    if dev.startswith('gpu') and theano.sandbox.cuda.dnn_available():
        pass
    elsif dev.startswith('cuda') and theano.gpuarray.dnn.dnn_available(dev):
        pass
    else: use_cudnn = False

This can be refactored, but I think it show more clearly the logic to use.
